File: android/src/main/java/com/reactnativeleanplum/RNLeanplum.java
package com.reactnativeleanplum;

import android.app.Application;
import android.content.pm.PackageManager;
import android.os.Bundle;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.ReadableType;

import com.leanplum.callbacks.StartCallback;
import com.leanplum.callbacks.VariablesChangedCallback;
import com.leanplum.Leanplum;
import com.leanplum.LeanplumActivityHelper;

import java.util.HashMap;

public class RNLeanplum extends ReactContextBaseJavaModule {
    private Application application;
    private String application_id;
    private String dev_key;
    private String prod_key;
    public RNLeanplum(ReactApplicationContext reactContext, Application app) {
        super(reactContext);

        Leanplum.setApplicationContext(app);
        LeanplumActivityHelper.enableLifecycleCallbacks(app);
        application = app;
        try {

            application_id = getReactApplicationContext().getApplicationInfo().metaData.getString("com.reactnativeleanplum.APP_ID","");
            dev_key = getReactApplicationContext().getApplicationInfo().metaData.getString("com.reactnativeleanplum.DEV_KEY","");
            prod_key = getReactApplicationContext().getApplicationInfo().metaData.getString("com.reactnativeleanplum.PROD_KEY","");

            // Insert your API keys here.
            if (BuildConfig.DEBUG) {
              Leanplum.setAppIdForDevelopmentMode(application_id, dev_key);
              Leanplum.enableTestMode();
            } else {
              Leanplum.setAppIdForProductionMode(application_id, prod_key);
            }
            Leanplum.start(app);
        } catch(Exception e){

        }
    }

    @Override
    public String getName() {
        return "Leanplum";
    }

    @ReactMethod
    public void setApiConnectionSettings(String hostName, String servletName, boolean ssl) {
        Leanplum.setApiConnectionSettings(hostName, servletName, ssl);
    }

    @ReactMethod
    public void setNetworkTimeout(int seconds, int downloadSeconds) {
        Leanplum.setNetworkTimeout(seconds, downloadSeconds);
    }
    
    @ReactMethod
    public void setCanDownloadContentMidSessionInProductionMode(boolean enabled) {
        Leanplum.setCanDownloadContentMidSessionInProductionMode(enabled);
    }
    
    @ReactMethod
    public void setFileHashingEnabledInDevelopmentMode(boolean enabled) {
        Leanplum.setFileHashingEnabledInDevelopmentMode(enabled);
    }
    
    
    @ReactMethod
    public void setAppIdForDevelopmentMode(String appId, String accessKey) {
        Leanplum.setAppIdForDevelopmentMode(appId, accessKey);
    }
    
    @ReactMethod
    public void setAppIdForProductionMode(String appId, String accessKey) {
        Leanplum.setAppIdForProductionMode(appId, accessKey);
    }
    
    @ReactMethod
    public void setDeviceId(String deviceId) {
        Leanplum.setDeviceId(deviceId);
    }
    
    @ReactMethod
    public void start(String userId, ReadableMap attributes, final Promise promise) {
        Leanplum.start(application, userId, attributes != null ? attributes.toHashMap() : null, new StartCallback() {
            @Override
            public void onResponse(boolean success) {
                promise.resolve(success);
            }
        });
    }
    
    @ReactMethod
    public void hasStarted(Promise promise) {
        promise.resolve(Leanplum.hasStarted());
    }

    @ReactMethod
    public void hasStartedAndRegisteredAsDeveloper(Promise promise) {
        promise.resolve(Leanplum.hasStartedAndRegisteredAsDeveloper());
    }
    
    @ReactMethod
    public void onStartResponse(final Callback callback) {
        Leanplum.addStartResponseHandler(new StartCallback() {
            @Override
            public void onResponse(boolean success) {
                callback.invoke(success);
            }
        });
    }
    
    @ReactMethod
    public void onVariablesChanged(final Callback callback) {
        Leanplum.addVariablesChangedHandler(new VariablesChangedCallback() {
            @Override
            public void variablesChanged() {
                callback.invoke();
            } 
        });
    }
    
    @ReactMethod
    public void onVariablesChangedAndNoDownloadsPending(final Callback callback) {
        Leanplum.addVariablesChangedAndNoDownloadsPendingHandler(new VariablesChangedCallback() {
            @Override
            public void variablesChanged() {
                callback.invoke();
            } 
        });
    }
    
    @ReactMethod
    public void onceVariablesChangedAndNoDownloadsPending(final Callback callback) {
        Leanplum.addOnceVariablesChangedAndNoDownloadsPendingHandler(new VariablesChangedCallback() {
            @Override
            public void variablesChanged() {
                callback.invoke();
            } 
        });
    }
    
    @ReactMethod
    public void setUserId(String userId) {
        Leanplum.setUserId(userId);
    }
    
    @ReactMethod
    public void setUserAttributes(String userId, ReadableMap attributes) {
        Leanplum.setUserAttributes(userId, attributes != null ? attributes.toHashMap() : null);
    }
    
    @ReactMethod
    public void setTrafficSourceInfo(ReadableMap info) {
        ReadableMapKeySetIterator iterator = info.keySetIterator();
        HashMap<String, String> map = new HashMap<String, String>();

        while (iterator.hasNextKey()) {
            String key = iterator.nextKey();

            if (info.getType(key) == ReadableType.String) {
                map.put(key, info.getString(key));
            }
        }

        Leanplum.setTrafficSourceInfo(map);
    }
    
    @ReactMethod
    public void advanceTo(String state, String info, ReadableMap params) {
        Leanplum.advanceTo(state, info, params != null ? params.toHashMap() : null);
    }
    
    @ReactMethod
    public void pauseState() {
        Leanplum.pauseState();
    }
    
    @ReactMethod
    public void resumeState() {
        Leanplum.resumeState();
    }
    
    @ReactMethod
    public void trackAllAppScreens() {
        Leanplum.trackAllAppScreens();
    }
    
    @ReactMethod
    public void trackPurchase(String event, double value, String currencyCode, ReadableMap params) {
        Leanplum.trackPurchase(event, value, currencyCode, params != null ? params.toHashMap() : null);
    }

    
    @ReactMethod
    public void track(String event, double value, String info, ReadableMap params) {
        Leanplum.track(event, value, info, params != null ? params.toHashMap() : null);
    }
    
    @ReactMethod
    public void variants(Promise promise) {
        promise.resolve(Leanplum.variants());
    }
    
    @ReactMethod
    public void forceContentUpdate(final Callback callback) {
        Leanplum.forceContentUpdate(new VariablesChangedCallback() {
            @Override
            public void variablesChanged() {
                callback.invoke();
            } 
        });
    }
    
    @ReactMethod
    public void enableTestMode() {
        Leanplum.enableTestMode();
    }
    
    @ReactMethod
    public void setTestModeEnabled(boolean isTestModeEnabled) {
        Leanplum.setIsTestModeEnabled(isTestModeEnabled);
    }
    
    @ReactMethod
    public void deviceId(Promise promise) {
        promise.resolve(Leanplum.getDeviceId());
    }
    
    @ReactMethod
    public void userId(Promise promise) {
        promise.resolve(Leanplum.getUserId());
    }
}

File: ios/RNLeanplum.m
//
//  RNLeanplum.m
//  RNLeanplum
//
//  Created by Bassel Dagher on 8/18/17.
//  Copyright © 2017-2018 Branding Brand. All rights reserved.
//

#import "RNLeanplum.h"

#if __has_include(<Leanplum/Leanplum.h>)
#import <Leanplum/Leanplum.h>
#elif __has_include("Leanplum.h")
#import "Leanplum.h"
#else
#import "Leanplum/Leanplum.h"
#endif

static NSString * const kRNLeanplumDuplicateStartErrorCode = @"RNLeanplumDuplicateStart";
static NSString * const kRNLeanplumDuplicateStartErrorReason = @"Leanplum: Already called start. Calling start a second time has no effect.";

@implementation RNLeanplum
RCT_EXPORT_MODULE(Leanplum);


RCT_REMAP_METHOD(setApiConnectionSettings,
                 setApiHostName:(NSString *)hostName withServletName:(NSString *)servletName usingSsl:(BOOL)ssl) {
    [Leanplum setApiHostName:hostName withServletName:servletName usingSsl:ssl];
}

RCT_REMAP_METHOD(setNetworkTimeout,
                 setNetworkTimeoutSeconds:(int)seconds forDownloads:(int)downloadSeconds) {
    [Leanplum setNetworkTimeoutSeconds:seconds forDownloads:downloadSeconds];
}

RCT_EXPORT_METHOD(setNetworkActivityIndicatorEnabled:(BOOL)enabled) {
    [Leanplum setNetworkActivityIndicatorEnabled:enabled];
}

RCT_EXPORT_METHOD(setCanDownloadContentMidSessionInProductionMode:(BOOL)value) {
    [Leanplum setCanDownloadContentMidSessionInProductionMode:value];
}

RCT_EXPORT_METHOD(setFileHashingEnabledInDevelopmentMode:(BOOL)enabled) {
    [Leanplum setFileHashingEnabledInDevelopmentMode:enabled];
}

RCT_EXPORT_METHOD(setVerboseLoggingInDevelopmentMode:(BOOL)enabled) {
    [Leanplum setVerboseLoggingInDevelopmentMode:enabled];
}

RCT_EXPORT_METHOD(setInAppPurchaseEventName:(NSString *)event) {
    [Leanplum setInAppPurchaseEventName:event];
}

RCT_REMAP_METHOD(setAppIdForDevelopmentMode,
                setAppId:(NSString *)appId withDevelopmentKey:(NSString *)accessKey) {
    [Leanplum setAppId:appId withDevelopmentKey:accessKey];
}

RCT_REMAP_METHOD(setAppIdForProductionMode,
                setAppId:(NSString *)appId withProductionKey:(NSString *)accessKey) {
    [Leanplum setAppId:appId withProductionKey:accessKey];
}

RCT_EXPORT_METHOD(setDeviceId:(NSString *)deviceId) {
    [Leanplum setDeviceId:deviceId];
}

RCT_EXPORT_METHOD(setAppVersion:(NSString *)appVersion) {
    [Leanplum setAppVersion:appVersion];
}

RCT_REMAP_METHOD(start,
                 startWithUserId:(NSString *)userId userAttributes:(NSDictionary *)attributes resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    if (![Leanplum hasStarted]) {
        return [Leanplum startWithUserId:userId userAttributes:attributes responseHandler:^(BOOL success) {
            return resolve(@(success));
        }];
    } else {
        reject(kRNLeanplumDuplicateStartErrorCode, kRNLeanplumDuplicateStartErrorReason, nil);
    }
}

RCT_REMAP_METHOD(hasStarted,
                hasStartedWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    return resolve(@([Leanplum hasStarted]));
}

RCT_REMAP_METHOD(hasStartedAndRegisteredAsDeveloper,
                hasStartedAndRegisteredAsDeveloperWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    return resolve(@([Leanplum hasStartedAndRegisteredAsDeveloper]));
}

RCT_EXPORT_METHOD(onStartResponse:(RCTResponseSenderBlock)callback) {
    return [Leanplum onStartResponse:^(BOOL success) {
        return callback(@[@(success)]);
    }];
}

RCT_EXPORT_METHOD(onVariablesChanged:(RCTResponseSenderBlock)callback) {
    return [Leanplum onVariablesChanged:^{
        return callback(@[]);
    }];
}

RCT_EXPORT_METHOD(onInterfaceChanged:(RCTResponseSenderBlock)callback) {
    return [Leanplum onInterfaceChanged:^{
        return callback(@[]);
    }];
}

RCT_EXPORT_METHOD(onVariablesChangedAndNoDownloadsPending:(RCTResponseSenderBlock)callback) {
    return [Leanplum onVariablesChangedAndNoDownloadsPending:^{
        return callback(@[]);
    }];
}

RCT_EXPORT_METHOD(onceVariablesChangedAndNoDownloadsPending:(RCTResponseSenderBlock)callback) {
    return [Leanplum onceVariablesChangedAndNoDownloadsPending:^{
        return callback(@[]);
    }];
}

RCT_EXPORT_METHOD(setUserId:(NSString *)userId) {
    return [Leanplum setUserId:userId];
}

RCT_REMAP_METHOD(setUserAttributes,
                 setUserId:(NSString *)userId withUserAttributes:(NSDictionary *)attributes) {
    return [Leanplum setUserId:userId withUserAttributes:attributes];
}

RCT_EXPORT_METHOD(setTrafficSourceInfo:(NSDictionary *)info) {
    return [Leanplum setTrafficSourceInfo:info];
}

RCT_EXPORT_METHOD(advanceTo:(NSString *)state withInfo:(NSString *)info andParameters:(NSDictionary *)params) {
    return [Leanplum advanceTo:state withInfo:info andParameters:params];
}

RCT_EXPORT_METHOD(pauseState) {
    return [Leanplum pauseState];
}

RCT_EXPORT_METHOD(resumeState) {
    return [Leanplum resumeState];
}

RCT_EXPORT_METHOD(trackAllAppScreens) {
    return [Leanplum trackAllAppScreens];
}

RCT_EXPORT_METHOD(trackPurchase:(NSString *)event withValue:(double)value andCurrencyCode:(NSString *)currencyCode andParameters:(NSDictionary *)params) {
    return [Leanplum trackPurchase:event withValue:value andCurrencyCode:currencyCode andParameters:params];
}

RCT_EXPORT_METHOD(trackInAppPurchases) {
    return [Leanplum trackInAppPurchases];
}

RCT_EXPORT_METHOD(track:(NSString *)event withValue:(double)value andInfo:(NSString *)info andParameters:(NSDictionary *)params) {
    return [Leanplum track:event withValue:value andInfo:info andParameters:params];
}

RCT_REMAP_METHOD(variants,
                 variantsWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    return resolve([Leanplum variants]);
}

RCT_EXPORT_METHOD(forceContentUpdate:(RCTResponseSenderBlock)callback) {
    if (callback) {
        return [Leanplum forceContentUpdate:^{
            return callback(@[]);
        }];
    }

    return [Leanplum forceContentUpdate];
}

RCT_EXPORT_METHOD(enableTestMode) {
    return [Leanplum enableTestMode];
}

RCT_EXPORT_METHOD(setTestModeEnabled:(BOOL)isTestModeEnabled) {
    return [Leanplum setTestModeEnabled:isTestModeEnabled];
}

RCT_EXPORT_METHOD(setPushSetup:(RCTResponseSenderBlock)callback) {
    return [Leanplum setPushSetup:^{
        return callback(@[]);
    }];
}

RCT_REMAP_METHOD(isPreLeanplumInstall,
                 isPreLeanplumInstallWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    return resolve(@([Leanplum isPreLeanplumInstall]));
}

RCT_REMAP_METHOD(deviceId,
                 deviceIdWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    return resolve([Leanplum deviceId]);
}

RCT_REMAP_METHOD(userId,
                 userIdWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    return resolve([Leanplum userId]);
}

@end
